File: 2025/04 April/16 Floyd Warshall/main.cpp
// Link: https://www.geeksforgeeks.org/problems/implementing-floyd-warshall2042/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    void floydWarshall(vector<vector<int>> &dist)
    {
        int V = dist.size();

        for (int k = 0; k < V; k++)
        {
            for (int i = 0; i < V; i++)
            {
                for (int j = 0; j < V; j++)
                {

                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }
};

int main()
{
    Solution obj;
    vector<vector<int>> dist = {
        {0, 4, (int)1e8, 5, (int)1e8},
        {(int)1e8, 0, 1, (int)1e8, 6},
        {2, (int)1e8, 0, 3, (int)1e8},
        {(int)1e8, (int)1e8, 1, 0, 2},
        {1, (int)1e8, (int)1e8, 4, 0}};

    obj.floydWarshall(dist);
    for (const auto &row : dist)
    {
        for (int val : row)
        {
            cout << val << " ";
        }
        cout << endl;
    }

    cout << endl;

    dist = {
        {0, -1, 2},
        {1, 0, (int)1e8},
        {3, 1, 0}};

    obj.floydWarshall(dist);
    for (const auto &row : dist)
    {
        for (int val : row)
        {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: 2025/04 April/16 Floyd Warshall/main.py
# Link: https://www.geeksforgeeks.org/problems/implementing-floyd-warshall2042/1


class Solution:
    def floydWarshall(self, dist):
        for k in range(len(dist)):
            for i in range(len(dist)):
                for j in range(len(dist)):
                    if dist[i][j] > dist[i][k] + dist[k][j]:
                        dist[i][j] = dist[i][k] + dist[k][j]
        return dist


if __name__ == "__main__":
    obj = Solution()
    dist = [
        [0, 4, 10**8, 5, 10**8],
        [10**8, 0, 1, 10**8, 6],
        [2, 10**8, 0, 3, 10**8],
        [10**8, 10**8, 1, 0, 2],
        [1, 10**8, 10**8, 4, 0],
    ]
    result = obj.floydWarshall(dist)
    for row in result:
        print(row)
    print()

    dist = [[0, -1, 2], [1, 0, 10**8], [3, 1, 0]]
    result = obj.floydWarshall(dist)
    for row in result:
        print(row)
